File: projectforge-business/src/main/java/org/projectforge/framework/persistence/database/DatabaseDao.java
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2019 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.framework.persistence.database;

import de.micromata.genome.jpa.StdRecord;
import org.apache.commons.lang3.ClassUtils;
import org.hibernate.ScrollMode;
import org.hibernate.ScrollableResults;
import org.hibernate.search.SearchFactory;
import org.hibernate.search.jpa.FullTextEntityManager;
import org.hibernate.search.jpa.Search;
import org.projectforge.business.timesheet.TimesheetDO;
import org.projectforge.framework.persistence.api.ExtendedBaseDO;
import org.projectforge.framework.persistence.api.ReindexSettings;
import org.projectforge.framework.persistence.entities.AbstractBaseDO;
import org.projectforge.framework.persistence.history.entities.PfHistoryMasterDO;
import org.projectforge.framework.persistence.jpa.PfEmgrFactory;
import org.projectforge.framework.time.DateHelper;
import org.projectforge.framework.time.DateTimeFormatter;
import org.projectforge.framework.time.DayHolder;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Repository;

import javax.persistence.EntityManager;
import javax.persistence.FlushModeType;
import javax.persistence.TypedQuery;
import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.From;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;

/**
 * Creates index creation script and re-indexes data-base.
 *
 * @author Kai Reinhard (k.reinhard@micromata.de)
 */
@Repository
public class DatabaseDao {
  private static final int MIN_REINDEX_ENTRIES_4_USE_SCROLL_MODE = 2000;

  private static final org.slf4j.Logger log = org.slf4j.LoggerFactory.getLogger(DatabaseDao.class);

  private Date currentReindexRun = null;

  @Autowired
  private PfEmgrFactory emgrFactory;

  /**
   * Since yesterday and 1,000 newest entries at maximimum.
   */
  public static ReindexSettings createReindexSettings(final boolean onlyNewest) {
    if (onlyNewest) {
      final DayHolder day = new DayHolder();
      day.add(Calendar.DAY_OF_MONTH, -1); // Since yesterday:
      return new ReindexSettings(day.getDate(), 1000); // Maximum 1,000 newest entries.
    } else {
      return new ReindexSettings();
    }
  }

  public <T> String rebuildDatabaseSearchIndices(final Class<T> clazz, final ReindexSettings settings) {
    if (currentReindexRun != null) {
      return "Another re-index job is already running. The job was started at: "
              + DateTimeFormatter.instance().getFormattedDateTime(currentReindexRun, Locale.ENGLISH, DateHelper.UTC)
              + " (UTC)";
    }
    final StringBuffer buf = new StringBuffer();
    reindex(clazz, settings, buf);
    return buf.toString();
  }

  public <T> void reindex(final Class<T> clazz, final ReindexSettings settings, final StringBuffer buf) {
    if (currentReindexRun != null) {
      buf.append(" (cancelled due to another running index-job)");
      return;
    }
    synchronized (this) {
      try {
        currentReindexRun = new Date();
        buf.append(ClassUtils.getShortClassName(clazz));
        reindex(clazz, settings);
        buf.append(", ");
      } finally {
        currentReindexRun = null;
      }
    }
  }

  /**
   * @param clazz
   */
  private <T> long reindex(final Class<T> clazz, final ReindexSettings settings) {
    if (settings.getLastNEntries() != null || settings.getFromDate() != null) {
      // OK, only partly re-index required:
      return reindexObjects(clazz, settings);
    }
    // OK, full re-index required:
    if (isIn(clazz, TimesheetDO.class, PfHistoryMasterDO.class)) {
      // MassIndexer throws LazyInitializationException for some classes, so use it only for the important classes (with most entries):
      return reindexMassIndexer(clazz);
    }
    return reindexObjects(clazz, null);
  }

  private boolean isIn(final Class<?> clazz, final Class<?>... classes) {
    for (final Class<?> cls : classes) {
      if (clazz.equals(cls)) {
        return true;
      }
    }
    return false;
  }

  private <T> long reindexObjects(final Class<T> clazz, final ReindexSettings settings) {
    return emgrFactory.runInTrans(emgr -> {
      final EntityManager em = emgr.getEntityManager();
      final Long number = getRowCount(em, clazz); // Get number of objects to re-index (select count(*) from).
      final boolean scrollMode = number > MIN_REINDEX_ENTRIES_4_USE_SCROLL_MODE;
      log.info("Starting re-indexing of "
              + number
              + " entries (total number) of type "
              + clazz.getName()
              + " with scrollMode="
              + scrollMode
              + "...");
      final int batchSize = 1000;// NumberUtils.createInteger(System.getProperty("hibernate.search.worker.batch_size")
      final FullTextEntityManager fullTextEntityManager = Search.getFullTextEntityManager(em);
      fullTextEntityManager.setFlushMode(FlushModeType.COMMIT);
      //HibernateCompatUtils.setFlushMode(fullTextSession, FlushMode.MANUAL);
      //HibernateCompatUtils.setCacheMode(fullTextSession, CacheMode.IGNORE);
      long index = 0;
      if (scrollMode) {
        // Scroll-able results will avoid loading too many objects in memory
        TypedQuery<T> query = createCriteria(em, clazz, settings);
        org.hibernate.query.Query hquery = query.unwrap(org.hibernate.query.Query.class);
        ScrollableResults results = hquery.scroll(ScrollMode.FORWARD_ONLY);
        while (results.next()) {
          final Object obj = results.get(0);
          if (obj instanceof ExtendedBaseDO<?>) {
            ((ExtendedBaseDO<?>) obj).recalculate();
          }
          fullTextEntityManager.index(obj);
          if (index++ % batchSize == 0) {
            fullTextEntityManager.flush(); // clear every batchSize since the queue is processed
          }
        }
      } else {
        TypedQuery<T> query = createCriteria(em, clazz, settings);
        final List<T> list = query.getResultList();
        for (final Object obj : list) {
          if (obj instanceof ExtendedBaseDO<?>) {
            ((ExtendedBaseDO<?>) obj).recalculate();
          }
          fullTextEntityManager.index(obj);
          if (index++ % batchSize == 0) {
            fullTextEntityManager.flush(); // clear every batchSize since the queue is processed
          }
        }
      }
      final SearchFactory searchFactory = fullTextEntityManager.getSearchFactory();
      searchFactory.optimize(clazz);
      log.info("Re-indexing of " + index + " objects of type " + clazz.getName() + " done.");
      return index;
    });
  }

  private <T> long reindexMassIndexer(final Class<T> clazz) {
    return emgrFactory.runInTrans(emgr -> {
      final EntityManager em = emgr.getEntityManager();
      final Long number = getRowCount(em, clazz); // Get number of objects to re-index (select count(*) from).
      log.info("Starting (mass) re-indexing of " + number + " entries of type " + clazz.getName() + "...");
      final FullTextEntityManager fullTextEntityManager = Search.getFullTextEntityManager(em);
      try {
        fullTextEntityManager.createIndexer(clazz)//
                .batchSizeToLoadObjects(25) //
                //.cacheMode(CacheMode.NORMAL) //
                .threadsToLoadObjects(5) //
                //.threadsForIndexWriter(1) //
                .startAndWait();
      } catch (final InterruptedException ex) {
        log.error("Exception encountered while reindexing: " + ex.getMessage(), ex);
      }
      final SearchFactory searchFactory = fullTextEntityManager.getSearchFactory();
      searchFactory.optimize(clazz);
      log.info("Re-indexing of " + number + " objects of type " + clazz.getName() + " done.");
      return number;
    });
  }

  private <T> Long getRowCount(final EntityManager entityManager, final Class<T> clazz) {
    final CriteriaBuilder cb = entityManager.getCriteriaBuilder();
    CriteriaQuery<Long> cq = cb.createQuery(Long.class);
    cq.select(cb.count(cq.from(clazz)));
    return entityManager.createQuery(cq).getSingleResult();
  }

  private <T> TypedQuery<T> createCriteria(final EntityManager entityManager, final Class<T> clazz, final ReindexSettings settings) {
    final CriteriaBuilder cb = entityManager.getCriteriaBuilder();
    CriteriaQuery<T> cr = cb.createQuery(clazz);
    From root = cr.from(clazz);
    if (settings != null) {
      String lastUpdateProperty = null;
      if (AbstractBaseDO.class.isAssignableFrom(clazz)) {
        lastUpdateProperty = "lastUpdate";
      } else if (StdRecord.class.isAssignableFrom(clazz)) {
        lastUpdateProperty = "modifiedAt";
      }
      if (lastUpdateProperty != null && settings.getFromDate() != null) {
        cb.equal(root.get(lastUpdateProperty), settings.getFromDate());
      }
      if (settings.getLastNEntries() != null) {
        if (clazz.isAssignableFrom(PfHistoryMasterDO.class)) {
          cr.orderBy(cb.desc(root.get("pk")));
        } else {
          cr.orderBy(cb.desc(root.get("id")));
        }
      }
    }
    TypedQuery<T> query = entityManager.createQuery(cr);
    if (settings != null && settings.getLastNEntries() != null) {
      query.setMaxResults(settings.getLastNEntries());
    }
    return query;
  }
}

File: projectforge-business/src/main/kotlin/org/projectforge/framework/persistence/database/DatabaseDao.kt
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2026 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.framework.persistence.database

import jakarta.persistence.EntityManager
import jakarta.persistence.EntityManagerFactory
import kotlinx.coroutines.future.await
import mu.KotlinLogging
import org.apache.commons.lang3.ClassUtils
import org.hibernate.search.mapper.orm.Search
import org.hibernate.search.mapper.orm.massindexing.MassIndexer
import org.hibernate.search.mapper.orm.session.SearchSession
import org.hibernate.search.mapper.pojo.massindexing.MassIndexingMonitor
import org.projectforge.framework.persistence.api.ReindexSettings
import org.projectforge.framework.time.DateHelper
import org.projectforge.framework.time.DateTimeFormatter
import org.projectforge.framework.time.DayHolder
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.*

private val log = KotlinLogging.logger {}

/**
 * Creates index creation script and re-indexes data-base.
 *
 * @author Kai Reinhard (k.reinhard@micromata.de)
 */
// Check open connections in PostgreSQL:
// SELECT backend_start, query_start, state_change, wait_event_type, state, query  FROM pg_stat_activity where state <> 'idle';
@Service
open class DatabaseDao {
    private var currentReindexRun: Date? = null

    @Autowired
    private lateinit var entityManagerFactory: EntityManagerFactory

    @JvmOverloads
    fun <T> rebuildDatabaseSearchIndices(clazz: Class<T>, settings: ReindexSettings = ReindexSettings()): String {
        if (currentReindexRun != null) {
            val otherJobStarted =
                DateTimeFormatter.instance().getFormattedDateTime(currentReindexRun, Locale.ENGLISH, DateHelper.UTC)
            return ("Another re-index job is already running. The job was started at: $otherJobStarted (UTC)")
        }
        val sb = StringBuilder()
        reindex(clazz, settings, sb)
        return sb.toString()
    }

    fun <T> reindex(clazz: Class<T>, settings: ReindexSettings, sb: StringBuilder) {
        if (currentReindexRun != null) {
            sb.append(" (cancelled due to another running index-job)")
            return
        }
        synchronized(this) {
            try {
                currentReindexRun = Date()
                sb.append(ClassUtils.getShortClassName(clazz))
                reindexObjects(clazz, settings)
                sb.append(", ")
            } finally {
                currentReindexRun = null
            }
        }
    }

    /**
     * Blocking re-index run, used by the classic clients (Wicket admin page, cron job, synchronous REST calls).
     * For a cancellable run inside a coroutine use [reindexSuspending].
     */
    private fun <T> reindexObjects(clazz: Class<T>, settings: ReindexSettings) {
        entityManagerFactory.createEntityManager().use { em ->
            try {
                createMassIndexer(em, clazz, settings, IndexProgressMonitor(clazz))
                    .startAndWait() // Blockiert, bis die Indizierung abgeschlossen ist
            } catch (ex: InterruptedException) {
                log.error(ex.message, ex)
            }
        }
    }

    /**
     * Re-indexes the given class without blocking the calling thread and, unlike [rebuildDatabaseSearchIndices],
     * abortable: [MassIndexer.startAndWait] can't be interrupted by a coroutine, while cancelling the future of
     * [MassIndexer.start] really stops the indexing (see CancellableExecutionCompletableFuture of Hibernate Search).
     *
     * Serializing concurrent runs is up to the caller (jobs do it via their queue strategy).
     */
    suspend fun <T> reindexSuspending(
        clazz: Class<T>,
        settings: ReindexSettings,
        monitor: MassIndexingMonitor = IndexProgressMonitor(clazz),
    ) {
        // The EntityManager has to stay open until the indexer is done, so await() happens inside use { }.
        entityManagerFactory.createEntityManager().use { em ->
            createMassIndexer(em, clazz, settings, monitor).start().await()
        }
    }

    private fun <T> createMassIndexer(
        em: EntityManager,
        clazz: Class<T>,
        settings: ReindexSettings,
        monitor: MassIndexingMonitor,
    ): MassIndexer {
        // totalEntries are given by Hibernate search to MassIndexingMonitor.
        val searchSession: SearchSession = Search.session(em)
        val indexer = searchSession.massIndexer(clazz)
            .threadsToLoadObjects(4) // Anzahl der Threads zum Laden von Entitäten
            .batchSizeToLoadObjects(25) // Batch-Größe
            .idFetchSize(150) // Größe des ID-Fetch
            .monitor(monitor) // Fortschrittsmonitor hinzufügen
        val fromDate = settings.fromDate
        val strategy = ReindexerRegistry.get(clazz)
        val modifiedAtProperty = strategy.modifiedAtProperty
        if (fromDate != null && modifiedAtProperty != null) {
            // Only the recently modified entries: the rest of the index has to survive, so no purge. Without this
            // (purgeAllOnStart defaults to true) a partial run would wipe every document not touched by it.
            indexer.purgeAllOnStart(false)
            val condition = StringBuilder("$modifiedAtProperty >= :fromDate")
            // The change history holds the rows of all entities, so a run started for a list page has to say whose
            // (otherwise re-indexing the book list would also re-index yesterday's history of every other entity).
            // Several names, because history rows are written per entity instance: the order list needs the history
            // of its positions and payment schedules, too. Never empty, see ReindexSettings.
            val entityNames = settings.entityNames?.takeIf { strategy.entityNameProperty != null }
            if (entityNames != null) {
                condition.append(" and ${strategy.entityNameProperty} in :entityNames")
            }
            log.info { "${clazz.simpleName}: Re-indexing only where $condition, fromDate=$fromDate, entityNames=$entityNames" }
            val step = indexer.type(clazz).reindexOnly(condition.toString()).param("fromDate", fromDate)
            // Hibernate ORM's setParameter detects the collection of an in-parameter and binds it as a list.
            entityNames?.let { step.param("entityNames", it) }
        } else if (fromDate != null) {
            log.info { "${clazz.simpleName}: No property of last modification known, so all entries are re-indexed." }
        }
        return indexer
    }

    companion object {
        /**
         * Since yesterday. [ReindexSettings.getLastNEntries] is set for the classic clients displaying it, but has no
         * effect on the indexing itself: reindexOnly of Hibernate Search takes a where condition without order or
         * limit, and limitIndexedObjectsTo would cap arbitrary entries, not the newest ones.
         *
         * @param entityNames The entities the run was started for (see BaseDao.historyEntityNames), restricting the
         *          change history to their rows. Null or empty for a system wide run, whose history isn't restricted.
         */
        @JvmStatic
        @JvmOverloads
        fun createReindexSettings(onlyNewest: Boolean, entityNames: Collection<String>? = null): ReindexSettings {
            return if (onlyNewest) {
                val day = DayHolder()
                day.add(Calendar.DAY_OF_MONTH, -1) // Since yesterday:
                ReindexSettings(day.utilDate, 1000, entityNames) // Maximum 1,000 newest entries.
            } else {
                ReindexSettings()
            }
        }
    }
}
